Site Settings: move Date Time Formats in Writing and enable it in production

Copons opened the pull request:
This PR moves the Date and Time Formats away from Settings → General and into Settings → Writing (Composing block).

It also enables it for everybody in .com and in Jetpack, provided they're at least on v4.7 (see https://github.com/Automattic/jetpack/commit/a9adf86c3f4b71a89536650fa9d52ca6d43d27f2: `* Changelog: add #6343`).

## New Version

![apr-07-2017 13-54-56](https://cloud.githubusercontent.com/assets/2070010/24800887/e5df9354-1b99-11e7-8696-96b2979ade47.gif)

## Old Version

![apr-07-2017 13-57-36](https://cloud.githubusercontent.com/assets/2070010/24800957/2fb1b138-1b9a-11e7-9173-54e7cac32c55.gif)

drw158 commented:
I like that it's now an expandable card. I like where it is located.

Can the whole row be clickable? How do the Jetpack settings with expandable cards behave?

Copons commented:
@drw158 
> Can the whole row be clickable?

Sure thing, is a matter of adding a prop to the foldable card. Push coming right up.

> How do the Jetpack settings with expandable cards behave?

Not sure I got what you mean, but this is a gif of the Writing settings for a Jetpack site:

![apr-07-2017 16-36-17](https://cloud.githubusercontent.com/assets/2070010/24807612/cc293d40-1bb0-11e7-8a9b-2de0084b299e.gif)

drw158 commented:
Cool! Let's make the whole row clickable then.

Copons commented:
@drw158 Forgot to push, now it's in. :)

tyxla commented on client/my-sites/site-settings/composing/index.jsx:
```diff
@@ -10,44 +10,63 @@ import { connect } from 'react-redux';
 import DefaultPostFormat from './default-post-format';
 import AfterTheDeadline from './after-the-deadline';
-import { isJetpackSite, siteSupportsJetpackSettingsUi } from 'state/sites/selectors';
+import DateTimeFormat from '../date-time-format';
```
I'd suggest avoiding relative paths, especially in site settings. Settings are known to be moved every now and then, so it's just easier if we import them by using absolute paths.

tyxla commented on client/my-sites/site-settings/form-general.jsx:
```diff
@@ -352,51 +348,6 @@ class SiteSettingsFormGeneral extends Component {
 		);
 	}
 
-	dateTimeFormat() {
```
In addition to removing date & time format from General settings, let's also remove:

* The `date-time-format` route from `client/my-sites/site-settings/index.js`
* `DateTimeFormat` and `dateTimeFormat` from `client/my-sites/site-settings/controller.js`

because they'll no longer be necessary as we manage these settings inline. And we don't want to include the date & time stuff into multiple chunks, right? 😉

tyxla commented on client/my-sites/site-settings/date-time-format/index.jsx:
```diff
@@ -1,39 +1,178 @@
+	phpToMomentDatetimeFormat,
+} from './utils';
+
+export class DateTimeFormat extends Component {
```
If you were starting with this PR right now, I'd recommend splitting this PR to two:

* Changing Date & Time to inline and removing unused code and route.
* Moving the inline Date & time to Writing.

This would reduce the change of introducing subtle issues, and would allow the PR to be reviewed easier. But it's your call really.

tyxla commented on client/my-sites/site-settings/form-general.jsx:
```diff
@@ -486,8 +437,6 @@ class SiteSettingsFormGeneral extends Component {
 					</form>
 				</Card>
 
-				{ this.dateTimeFormat() }
```
Some further cleanup of `form-general` would be recommended here:

* Removing unnecessary fields from `defaultSettings` and `formSettings`, for example `date_format`, `time_format` and `start_of_week`.
*  Removing the GMT offset logic from the `wrapSettingsForm` call.

Copons commented:
@tyxla You're definitely correct. Somehow I forgot about the routes I added for the sub-page (and, well, for the settings left in form-general, I relied too much on ESLint unused variables which of course doesn't catch unused properties).

I'd fix those in this PR to get it merged as soon as possible actually. :)

Copons commented:
@tyxla Removed the route from `index` and `controller`; removed the settings from `form-general` (I've left `gmt_offset` though, as it's used by the Timezone setting).

---

@drw158 I've changed the foldable card title style a little bit to make it look similar as the "Default Post Format" label: bold and with a tiny bottom margin.
The reasoning is that without the bold the flow of the Composing card looked a bit off:
[label] → [setting] → [label]
whereas with both settings in bold is more clear that it's:
[label] → [setting] → [setting]

| Before | After |
| --- | --- |
| <img width="668" alt="screen shot 2017-04-10 at 12 08 23" src="https://cloud.githubusercontent.com/assets/2070010/24859082/f3abf8c0-1de6-11e7-8c27-0aadaadc9175.png"> | <img width="667" alt="screen shot 2017-04-10 at 12 08 10" src="https://cloud.githubusercontent.com/assets/2070010/24859084/f7af9878-1de6-11e7-8cb6-d5cf95393cf6.png"> |

Copons commented:
@tyxla I've removed the leftover stuff and rebase after your change on the ATD card. It seems everything is working, but just wanted to give you a heads up on it in case I've missed something (there were conflicts on 2 different commits, so it got quite confusing)

tyxla commented:
@Copons thanks for the heads up, will look again in a bit :+1:

tyxla commented on client/my-sites/site-settings/form-writing.jsx:
```diff
@@ -161,39 +165,60 @@ const connectComponent = connect(
+	] );
+
+	// handling `gmt_offset` and `timezone_string` values
+	const gmt_offset = settings.gmt_offset;
```
The same logic is currently being duplicated in the general form and the writing form. May I suggest:
* Taking this logic out in a separate utility method.
* Adding tests for that method

tyxla commented on client/my-sites/site-settings/date-time-format/index.jsx:
```diff
@@ -1,39 +1,178 @@
+
+export class DateTimeFormat extends Component {
+	static propTypes = {
+		handleSelect: PropTypes.func,
```
`handleSelect` should be a required prop, right? The same applies to `updateFields`, which is also missing from the list of `propTypes`.

Copons commented:
@tyxla If it's ok with you, I'll add the two `isRequired` and merge without further review (those are required props in the parent component anyway).
I'd leave the `gmt_offset` part for another time, though, since it mostly touches the Timezone setting.

tyxla commented:
Sure, go ahead 👍  The `gmt_offset` part can always be iterated on later.
